File: Task993/Sources/Task993/validation.swift
import Foundation
#if DEVELOPING
import TestData

let data1 = TestData(enabled: true, name: "Data 1", text: """
4 3
1 3 6
0 1 11
1 2 7
""")

let data2 = TestData(enabled: true, name: "Data 2", text: """
5 10
3 1 46
2 1 48
4 2 21
0 1 26
4 1 50
4 0 87
1 0 57
1 3 42
1 2 52
0 4 10
""")

let data3 = TestData(enabled: false, name: "Data 3", text: """
""")

let data4 = TestData(enabled: false, name: "Data 4", text: """
""")

let data5 = TestData(enabled: false, name: "Data 5", text: """
""")

let validationList = [data1, data2, data3, data4, data5]

#endif

File: Task1381Q36X/src/Validation.kt
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.PrintStream

fun main() {
    val data1 = TestData("Data 1", """
        8 8
        ..#.....
        .###.##.
        .###..#.
        ####....
        .#..###.
        .....###
        ....##.#
        .....##.
    """.trimIndent(), """
        3
    """.trimIndent())

    val data2 = TestData("Data 2", """
        3 5
        #.#.#
        .###.
        #.#.#
    """.trimIndent(), """
        5
    """.trimIndent())

    val data3 = TestData("Data 3", """
        5 2
        ..
        ..
        ..
        ..
        ..
    """.trimIndent(), """
        0
    """.trimIndent())

    val data4 = TestData("Data 4", """
    """.trimIndent(), """
    """.trimIndent())

    val data5 = TestData("Data 5", """
    """.trimIndent(), """
    """.trimIndent())

    val validationList = listOf(
        data1,
        data2,
        data3,
//        data4,
//        data5,
    )
    
    println()
    validationList.forEach { it.execute(::solve) }
    println("--------------------------------")
    println()
}

class TestData(
    private val name: String,
    private val data: String,
    private val expected: String,
) {
    fun execute(block: () -> Unit) {
        println("--- [$name] -------------------")
        val originalIn = System.`in`
        val originalOut = System.out
        val actual = ByteArrayInputStream(data.toByteArray()).use { inputStream ->
            ByteArrayOutputStream().use { outputStream ->
                System.setIn(inputStream)
                try {
                    System.setOut(PrintStream(outputStream))
                    try {
                        block()
                        outputStream.toString()
                    } finally {
                        System.setOut(originalOut)
                    }
                } finally {
                    System.setIn(originalIn)
                }
            }
        }
        print(actual)
        if (actual[actual.length - 1] != '\n') {
            println()
        }

        if (expected.isNotEmpty()) {        
            if (actual.trim() == expected.trim()) {
                println("✅ OK")
            } else {
                println("❌ NG")
            }
        }
    }
}
